Drop support for Python 3.8

jacobtylerwalls opened the pull request:
Remove support for launching pylint with Python 3.8.
Code that supports Python 3.8 can still be linted with the ``--py-version=3.8`` setting.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

Pierre-Sassoulas commented on pylint/checkers/method_args.py:
```diff
@@ -41,7 +41,6 @@ class MethodArgsChecker(BaseChecker):
             "positional-only-arguments-expected",
             "Emitted when positional-only arguments have been passed as keyword arguments. "
             "Remove the keywords for the affected arguments in the function call.",
-            {"minversion": (3, 8)},
```
Shouldn't we replace this with a py-version check ?

jacobtylerwalls replied:
Hmm, but if my library supports 3.7 - 3.12, and I'm using a 3.8 function with positional-only arguments in the wrong manner, wouldn't I rather continue to see the message even if `--py-version=3.7`, because I certainly have bigger problems anyway if I'm using a 3.8+ function when I supposedly support 3.7?

jacobtylerwalls replied:
I thought `--py-version` was most helpful for ensuring we silence suggestion-style messages that don't make sense on lower versions, like `consider-using-f-string`, but here we have a message that won't ever get raised on a lower version anyway, so I don't think we need to add a `--py-version` workaround for it.

DanielNoord replied:
I agree with Jacob's second comment. I don't think it will add much benefit to spend the time to add support for that edge case. Removing this line is fine.

Pierre-Sassoulas commented on pylint/message/message_definition_store.py:
```diff
@@ -58,9 +58,7 @@ def register_message(self, message: MessageDefinition) -> None:
-    @functools.lru_cache(  # pylint: disable=method-cache-max-size-none # noqa: B019
-        maxsize=None
-    )
+    @cache  # pylint: disable=method-cache-max-size-none # noqa: B019
```
Should we upgrade the comment above ? Do we have a false negative for cache ? I'm not well versed on the difference between cache and lru_cache.

jacobtylerwalls replied:
`@cache` == `@lru_cache(maxsize=None)`. So I don't think there's anything else to change here.

The pylint message clarifies:
```
 W1518: 'lru_cache(maxsize=None)' or 'cache' will keep all method args alive ...
```

Pierre-Sassoulas reviewed (commented):
Looks really great. I'm not sure we should remove the functional tests for py38/py39 (unless it's actual max_pyver=3.8) ?

jacobtylerwalls commented:
> I'm not sure we should remove the functional tests for py38/py39 (unless it's actual max_pyver=3.8) ?

I may not be following exactly, but it's worth remembering this: max_pyver=3.9 means run on 3.8 and below. I'd rather that be what happens when you put max_pyver=3.8, but it is what it is 🤷 . So anything with max_pyver=3.9 won't run when 3.9 is the minimum version and can be deleted, is my understanding.

Pierre-Sassoulas commented:
> I may not be following exactly, but it's worth remembering this: max_pyver=3.9 means run on 3.8 and below. I'd rather that be what happens when you put max_pyver=3.8, but it is what it is 🤷 . 

Yeah, didn't have that in mind either 😅 I'm feeling like making a breaking change, it's likely that (mainly) only us pylint maintainers use that atm. I never saw a plugin with functional tests in the wild. I'll re-review with that in mind later.

DanielNoord commented on .github/workflows/tests.yaml:
```diff
@@ -269,7 +269,7 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: ["pypy-3.8", "pypy-3.9"]
+        python-version: ["pypy-3.9"]
```
We should probably look into adding `pypy3.10`

jacobtylerwalls replied:
Surprise, it works! #9777

DanielNoord commented on pylint/testutils/output_line.py:
```diff
@@ -41,11 +39,11 @@ class OutputLine(NamedTuple):
         column = cls._get_column(msg.column)
-        end_line = cls._get_py38_none_value(msg.end_line, check_endline)
-        end_column = cls._get_py38_none_value(msg.end_column, check_endline)
+        end_line = msg.end_line
```
Let's just pass these directly?

DanielNoord reviewed (commented):
Awesome job!

Just one comment from me.

There are some functional tests where it would have been nice to convert them to be messages when `py-version==3.7` but that is a lot of work and I don't think is something we should focus our limited time on.

DanielNoord reviewed (approved):
:rocket:

jacobtylerwalls commented:
Last push resolves merge conflict with #9777

jacobtylerwalls commented:
Latest push to resolve merge conflict

Pierre-Sassoulas reviewed (commented):
Sorry I wasn't able to review faster. 

I feel like we're loosing a lot of valuable interpreter specific knowledge that could be integrated in py-version when we remove functional tests that ran only for EOL interpreters without looking in detail at what's happening. We already did that in the past but I think it's making``py-version`` kinda weak and 3.8 is easier to support than 3.7 because the ast's column/line is so much closer with other interpreters. I don't know if you looked at each test in detail and cleaned up the copy pastes or simple removed the one that ran only on EOL interpreters, Jacob ? There's a lot to review there I still didn't do it, but I feel like, for example, the carefully crafted tests about typing in python 3.8 from ``g/generic_alias/generic_alias_collections_py37.py`` brought a lot of value and the removal is going to hurt when someone open issues postfaced "I use py-version 3.8, pylint should not suggest impossible things for my interpreter, it was working fine in *old version*".

jacobtylerwalls commented:
Thanks for the review Pierre, and sorry for rushing you.

> I don't know if you looked at each test in detail and cleaned up the copy pastes or simple removed the one that ran only on EOL interpreters, Jacob ?

I just removed tests that ran only on EOL interpreters. If there are tests showing functionality that we can capture with py-version, I'm more than happy to make a new issue and itemize each one showing whether it's been investigated. It feels more "useful and optional" than dropping python 3.9, which I think we needed to do this summer, and I had a day to do it.

> There are some functional tests where it would have been nice to convert them to be messages when py-version==3.7 but that is a lot of work and I don't think is something we should focus our limited time on.

This is where I was at too, but as I mentioned above I'm happy to stub out an issue to investigate. I have separate commits in this PR to make this easier. However...

> for example, the carefully crafted tests about typing in python 3.8 from g/generic_alias/generic_alias_collections_py37.py brought a lot of value 

I'm not sure I see what we can salvage here. How are we going to know to raise `unsubscriptable-object` when `py-version` is 3.8 or below? Won't that require astroid changes? We've never done that to my knowledge, but my imagination might be a little weak here :D

jacobtylerwalls commented on doc/development_guide/contributor_guide/tests/writing_test.rst:
```diff
@@ -61,7 +61,6 @@ test runner. The following options are currently supported:
 
 - "min_pyver": Minimal python version required to run the test
 - "max_pyver": Python version from which the test won't be run. If the last supported version is 3.9 this setting should be set to 3.10.
-- "min_pyver_end_position": Minimal python version required to check the end_line and end_column attributes of the message
```
Thanks for the explanation, will revert.

jacobtylerwalls commented on pylint/testutils/lint_module_test.py:
```diff
@@ -81,9 +79,6 @@ def __init__(
         self._linter._arg_parser.add_argument(
             "--max_pyver", type=parse_python_version, default=(4, 0)
         )
-        self._linter._arg_parser.add_argument(
-            "--min_pyver_end_position", type=parse_python_version, default=(3, 8)
-        )
```
Leaving this as 3.8 for now because it's strongly typed as a tuple.

jacobtylerwalls commented:
@Pierre-Sassoulas we're approaching the season for dropping 3.9. I don't want to rush anyone this year. Any advice for how to proceed this time?

I think my view is still that we shouldn't block dropping tests on investigations about `--py-version` improvements, those can always be investigated by an interested party from the git history, but it's not an essential task (and is probably very expensive, requiring sophisticated astroid changes in some cases).

Pierre-Sassoulas commented:
Maybe we could only keep what does not require expensive astroid changes ? I'm just answering this from intuition, I didn't look into the details.

jacobtylerwalls commented:
Even the non-astroid changes are going to require engineering time as Daniël says here:

> There are some functional tests where it would have been nice to convert them to be messages when py-version==3.7 but that is a lot of work and I don't think is something we should focus our limited time on.

We can always do that later, but I just don't think we should block this cleanup work unless someone volunteers. I guess we can put out a call for volunteers?

Pierre-Sassoulas commented:
Isn't the time when we cleanup the easier tome to add a py-version rc file instead of removing ? I feel if we don't think about the maintenance cost when removing the code, we're never going to do it and the person that might want to do it will assume the maintenance cost was high. But we can choose to be honest about the fact that we're not going to support old python interpreter even if the cost is tiny (tiny as in "copy pasting the x.py file, renaming it x.rc and copy pasting py-version=3.9 inside x.rc, maybe changing some sys.version_info check to py-version checks inside the code if required or dropping both x.py and x.rc if the ast changed too much and making the check works would require an astroid update", or am I missing something ?). I suppose that I will be sad to see some typing functional tests be deleted again for 3.9, I remember that Marc did a great job with those at the time.

jacobtylerwalls commented:
I'll do a dry run of one file to see if it's as easy as that. 👍

cdce8p commented:
Something like #10382 might help with the `py-version` check issue.

jacobtylerwalls commented:
> I'll do a dry run of one file

I chose `alternative_union_syntax_error.py`, and it wasn't too bad, it already had `py_version`. I found that keeping the test file and merely doing this:

```diff
diff --git a/tests/functional/a/alternative/alternative_union_syntax_error.rc b/tests/functional/a/alternative/alternative_union_syntax_error.rc
index 5dc39b1a6..6f046622f 100644
--- a/tests/functional/a/alternative/alternative_union_syntax_error.rc
+++ b/tests/functional/a/alternative/alternative_union_syntax_error.rc
@@ -1,5 +1,2 @@
 [main]
 py-version=3.8
-
-[testoptions]
-max_pyver=3.10
```

Only leads to one new failure that I can't track down:

```py
E       AssertionError: Wrong message(s) raised for "alternative_union_syntax_error.py":
E       
E       Expected in testdata:
E         26: unsupported-binary-operation (times 2)
```

So I suppose in this case we can just comment out that test with a comment about a false negative and push onward?

***

There are only 8 files this time, and 4 of them are about deprecated methods. It's probably a much smaller lift to preserve these tests this time.
